### PR TITLE
Feature: k0s runtime remain work

### PR DESCRIPTION
**Pull request by starComingup**
### Describe what this PR does / why we need it
After k0s init step, there are update/join/delete/reset interface which need to implement, this PR finish all k0s cluster install action.

### Does this pull request fix one issue?
NONE, but a step for #1399

### CODE REVIEW

File: pkg/runtime/k0s/delete_masters.go
Comment (justadogistaken) [inline pkg/runtime/k0s/delete_masters.go:35]: I'd rather put those command at the caller place. This will break the readability.

File: pkg/runtime/k0s/delete_masters.go
Comment (justadogistaken) [inline pkg/runtime/k0s/delete_masters.go:46]: I guess there is no need to declare the temp var again.

File: pkg/runtime/k0s/delete_masters.go
Comment (allencloud) [inline pkg/runtime/k0s/delete_masters.go:34]: Please do not use format in the const string value. It would make code quite complicated.

File: pkg/runtime/k0s/delete_masters.go
Comment (allencloud) [inline pkg/runtime/k0s/delete_masters.go:35]: It should use an API calling torwards apiserver, rather than using a shell command.

File: pkg/runtime/k0s/delete_masters.go
Comment (allencloud) [inline pkg/runtime/k0s/delete_masters.go:49]: You should `return err` here, otherwise the eg.Go() could not catch the detailed error, and the `return eg.Wait()` would always `return nil`, which is definitely unreasonable.

File: pkg/runtime/k0s/delete_masters.go
Comment (allencloud) [inline pkg/runtime/k0s/delete_masters.go:62]: I think your error log message should be optimized. When logging error message, you should set up a rule for yourself, where to locate the wrap message. In this case, the wrap message is `failed to delete master:`.  And in Line 49, you have code `logrus.Errorf("failed to delete master %s: %v", master, err)`. Then if an error is returned back, then you error message would be `failed to delete master 1.1.1.1: failed to delete master: xxxxxx`. It is duplicated and definitely inproper.

I wish that you could pay more attention on every part of the code. Only by patience and carefulness, could quality of software be achieved.

File: pkg/runtime/k0s/delete_masters.go
Comment (allencloud) [inline pkg/runtime/k0s/delete_masters.go:59]: This is a detailed deep implementation of delete a master node. I think we should explain clearly what is the **STEPS** to delete a master by comments. If we understand the steps and then the code review could be quite easy. Actually we should review the design firstly, and code secondly.  No design there, and we can hardly do the code review. Do you think so ?

File: pkg/runtime/k0s/delete_masters.go
Comment (allencloud) [inline pkg/runtime/k0s/delete_masters.go:101]: Please use k8s apiClient to get the information you want. Otherwise it is quite strange to add shell action in golang code. If KubeCtl released a higher version, and the output format changes, the code here will immediately meet the compatibility issue. What a bad design. Right?

**Comment by allencloud** [comment]
I just reviewed the `delete_master.go` file, and attached several feedbacks. I think you have the same issue in other files, like `delete_node.go`, `join_node.go` and so on. Please update the original code all all files carefully. Then you could comment this pr to seek my further review. Thanks. @starComingup 🌹 🐯 🦁

File: pkg/runtime/k0s/delete_masters.go
Comment (starComingup) [reply pkg/runtime/k0s/delete_masters.go:59]: Yep! when I read kubernetes runtime module, I have the same feel. which files to delete? and what logic to follow. I so sorry to made this, fix this ASAP.

File: pkg/runtime/k0s/delete_masters.go
Comment (starComingup) [reply pkg/runtime/k0s/delete_masters.go:35]: Got it! next commit will fix it!

File: pkg/runtime/k0s/delete_masters.go
Comment (starComingup) [reply pkg/runtime/k0s/delete_masters.go:61]: I think the error will be thrown by ssh.Async and captured by the caller.

File: pkg/runtime/k0s/delete_masters.go
Comment (starComingup) [reply pkg/runtime/k0s/delete_masters.go:80]: I read this code block and I think It did a thing that judge the rest master nodes need greater than one, you code is more clearly and simplify

Comment (starComingup) [reply pkg/runtime/k0s/delete_masters.go:80]: As for `len(k.cluster.GetMasterIPList()) == 1)` ,i think it is no need to judge it.   Code logic will protect the cluster have a master at least.

File: apply/processor/utils.go
Comment (allencloud) [inline apply/processor/utils.go:25]: @starComingup Here I have a point to discuss with you. 

If runtime is a field of details processor instance, or belongs to a processor's function?

If runtime is a field of processor, then RuntimeChoose should be used in a processor's initialization, rather than set runtime in processor's function.

For example, I see 
```
type ScaleProcessor struct {
    runtime runtime.Interface
    xxx
}
```

And I see 
```
func (s *ScaleProcessor) PreProcess(cluster *v2.Cluster) error {
	runTime, err := RuntimeChoose(platform.DefaultMountClusterImageDir(cluster.Name), cluster, s.ClusterFile.GetKubeadmConfig())
	if err != nil {
		return fmt.Errorf("failed to init default runtime: %v", err)
	}
	s.Runtime = runTime
```
I feel quite strange. Why do you `RuntimeChoose` in ScaleProcessor and set ScaleProcessor's runtime filed, instead of set runtime type in `NewScaleProcessor`? Initialization in NewScaleProcessor could make code much more clear and more object-oriented. 

WDYT?

Comment (allencloud) [reply apply/processor/utils.go:25]: In addition, please change the function name into `ChooseRuntime` or `GetRuntime` rather than `RuntimeChoose`.

File: apply/processor/utils.go
Comment (allencloud) [inline apply/processor/utils.go:37]: I wish to see more thinking details about the `default` case. From my point of view, if there is no metadata.ClusterRuntime in image rootfs details, we take it as k8s by default. Because in the original implementation of ClusterImage, there is no metadata.ClusterRuntime, and we have to make the new version of sealer compatible with the legacy ClusterImage format. For what is worth, I suggest to add a log to tell end-user that he is using a legacy version, and using new version of sealer and ClusterImage is encouraged.

Otherwise, I think you should add a line of explicit log at Line31 to show the ClusterRuntime info.

File: pkg/runtime/k0s/reset.go
Comment (allencloud) [inline pkg/runtime/k0s/reset.go:50]: Do we need to use a `eg, _ := errgroup.WithContext(context.Background())` in `resetMasters` as well, just like what you did in `resetNodes`? @starComingup

File: pkg/runtime/k0s/runtime.go
Comment (allencloud) [inline pkg/runtime/k0s/runtime.go:44]: I have not see any assignment to this variable. Please double check it.

File: pkg/runtime/k0s/upgrade.go
Comment (allencloud) [inline pkg/runtime/k0s/upgrade.go:81]: Do we need to use a eg, _ := errgroup.WithContext(context.Background()) here?

File: pkg/runtime/k0s/upgrade.go
Comment (allencloud) [inline pkg/runtime/k0s/upgrade.go:50]: Do we need to use a `eg, _ := errgroup.WithContext(context.Background())` here?

**Comment by allencloud** [comment]
Left some comment, and wait for @starComingup 's feedback and update before merging.

File: apply/processor/utils.go
Comment (starComingup) [reply apply/processor/utils.go:37]: I agree with add a cluster runtime info and hint end-user using new version ClusterImage.

File: apply/processor/utils.go
Comment (starComingup) [reply apply/processor/utils.go:25]: > Here I have a point to discuss with you.
> 
> If runtime is a field of details processor instance, or belongs to a processor's function?
> 
> If runtime is a field of processor, then RuntimeChoose should be used in a processor's initialization, rather than set runtime in processor's function.
> 
> For example, I see

The old code set kubernetes.runtime here, so I just make this. But when I coded this, I also felt confused why origin logic set runtime in sub function. I will try to modify it!

File: pkg/runtime/k0s/reset.go
Comment (starComingup) [reply pkg/runtime/k0s/reset.go:50]: I think we can do what you advice, then I will verify it.

File: pkg/runtime/k0s/runtime.go
Comment (starComingup) [reply pkg/runtime/k0s/runtime.go:44]: Using this in Line45

File: pkg/runtime/k0s/upgrade.go
Comment (starComingup) [reply pkg/runtime/k0s/upgrade.go:81]: I don't think it can speed up too much, so I'll keep this logic for now.

File: pkg/runtime/k0s/upgrade.go
Comment (starComingup) [reply pkg/runtime/k0s/upgrade.go:50]: keep this logic for now, If necessary, I will modify it.

**Comment by starComingup** [comment]
All comment and code updated! Waiting for test.

**Review by allencloud — APPROVED** [review]

**Comment by allencloud** [comment]
Thanks greatly for your patience during review. @starComingup I think your work could improve sealer quite a lot. Thanks again.

**Pull request merged**
